sync/src/orphan_block_pool.rs

```diff
@@ -1,152 +1,260 @@
+    pub fn with_capacity(capacity: usize) -> Self {
+        OrphanBlockPool {
+            inner: RwLock::new(InnerOrphanBlockPool::with_capacity(capacity)),
         }
-
-        debug!("orphan pool pop chain len: {}", removed.len());
-        debug_assert_ne!(
-            removed.len(),
-            0,
-            "orphan pool removed list must not be zero"
-        );
-
-        shrink_to_fit!(self.blocks, SHRINK_THRESHOLD);
-        shrink_to_fit!(self.parents, SHRINK_THRESHOLD);
-        shrink_to_fit!(self.leaders, SHRINK_THRESHOLD);
-        removed
+    }
+    pub fn insert(&self, block: core::BlockView) -> Result<(), Error> {
+        self.inner.write().insert(block)
+    }
+    pub fn remove_blocks_by_parent(&self, root: &InputHash) -> Option<Vec<core::BlockView>> {
+        self.inner.write().remove_blocks_by_parent(root)
+    }
+    pub fn get_block(&self, hash: &InputHash) -> Option<core::BlockView> {
+        self.inner.read().get_block(hash)
+    }
+    pub fn is_empty(&self) -> bool {
+        self.inner.read().is_trees_empty() && self.inner.read().is_rootof_empty()
+    }
+    pub fn len(&self) -> usize {
+        self.inner.read().trees_len()
     }
 
-    pub fn get_block(&self, hash: &packed::Byte32) -> Option<core::BlockView> {
-        self.parents.get(hash).and_then(|parent_hash| {
-            self.blocks
-                .get(&parent_hash)
-                .and_then(|blocks| blocks.get(hash).cloned())
-        })
+    #[allow(dead_code)]
+    fn trees_len(&self) -> usize {
+        self.inner.read().trees_len()
+    }
+    #[allow(dead_code)]
+    fn root_of_len(&self) -> usize {
+        self.inner.read().rootof_len()
     }
 }
 
-// NOTE: Never use `LruCache` as container. We have to ensure synchronizing between
-// orphan_block_pool and block_status_map, but `LruCache` would prune old items implicitly.
-// RwLock ensures the consistency between maps. Using multiple concurrent maps does not work here.
-#[derive(Default)]
-pub struct OrphanBlockPool {
-    inner: RwLock<InnerPool>,
+pub struct InnerOrphanBlockPool {
+    // store root and sub-node block
+    trees: HashMap<ParentHash, VecDeque<(Hash, core::BlockView)>>,
+    // CP(child-parent) relationship between sub-nodes
+    root_of: HashMap<Hash, ParentHash>,
 }
 
-impl OrphanBlockPool {
+fn internal_error(reason: String) -> Error {
+    InternalErrorKind::System.other(reason).into()
+}
+
+impl InnerOrphanBlockPool {
     pub fn with_capacity(capacity: usize) -> Self {
-        OrphanBlockPool {
-            inner: RwLock::new(InnerPool::with_capacity(capacity)),
+        InnerOrphanBlockPool {
+            trees: HashMap::with_capacity(capacity),
+            root_of: HashMap::new(),
         }
     }
 
-    /// Insert orphaned block, for which we have already requested its parent block
-    pub fn insert(&self, block: core::BlockView) {
-        self.inner.write().insert(block);
+    /// root_of table contains direct child-parent relationship  
```

Comment:
```suggestion
    /// root_of table contains direct child-parent relationship
```

Lots of trailing whitespaces, not only in this line.